### PR TITLE
Scala Package, finish PrefetchIterator

### PR DESCRIPTION
**Pull request by Ldpe2G**
#1591

### CODE REVIEW

File: scala-package/core/src/test/scala/ml/dmlc/mxnet/IOSuite.scala [hunk 1]
```diff
@@ -150,6 +150,39 @@ class IOSuite extends FunSuite with BeforeAndAfterAll {
...
+    }
+
+    prefetchIter.reset()
+    while(prefetchIter.hasNext) {
+      prefetchIter.next()
+    }
+
+    prefetchIter.dispose()
+
+    assert(true)
```
Comment (javelinjs) [inline scala-package/core/src/test/scala/ml/dmlc/mxnet/IOSuite.scala]: Cool. Could you make some real check? e.g., batchCount

Comment (Ldpe2G) [inline scala-package/core/src/test/scala/ml/dmlc/mxnet/IOSuite.scala]: ok

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/io/PrefetchingIter.scala [hunk 2]
```diff
@@ -2,62 +2,162 @@ package ml.dmlc.mxnet.io
...
+    override def run(): Unit = {
+      while (started) {
+        dataTaken(i).acquire()
+        if (started) {
+          try {
+            nextBatch(i) = iters(i).next()
+          } catch {
+            case ex: NoSuchElementException => nextBatch(i) = null
+          }
+          dataReady(i).release()
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/io/PrefetchingIter.scala]: I think it is better to put this line outside of the if condition.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/io/PrefetchingIter.scala [hunk 3]
```diff
@@ -2,62 +2,162 @@ package ml.dmlc.mxnet.io
...
+    }
+  }
+
+  private val _batchSize: Int = this._provideData.toList(0)._2(0)
+  private val dataReady: IndexedSeq[Semaphore] =
+                                        (0 until iters.length).map(i => new Semaphore(0))
+  private val dataTaken: IndexedSeq[Semaphore] =
+                                        (0 until iters.length).map(i => new Semaphore(1))
+
+  private var started: Boolean = true
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/io/PrefetchingIter.scala]: mark as `@volatile`

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/io/PrefetchingIter.scala [hunk 4]
```diff
@@ -2,62 +2,162 @@ package ml.dmlc.mxnet.io
 
 import ml.dmlc.mxnet.{DataBatch, DataIter, NDArray, Shape}
 import org.slf4j.LoggerFactory
+import java.util.concurrent.Semaphore
 
 /**
  * TODO
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/io/PrefetchingIter.scala]: remove the TODO since you've implemented it :)

**Pull request merged**